### PR TITLE
Dotnet test generation

### PR DESCRIPTION
**Pull request by amidgol**

### CODE REVIEW

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace EvoMaster.Client.Util
+{
+    public class JsonInspector
+    {
+        public static bool ContainsKeyValue<T>(string json, string key, T value)
```
Comment (arcuri82) [inline client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs:11]: provide some documentation for this function

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            if (string.IsNullOrWhiteSpace(key))
+                throw new ArgumentNullException("Title");
```
Comment (arcuri82) [inline client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs:14]: why "Title"???

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+                matches = GetMatches(json, key);
+            }
+
+            return matches.Any(x => x.Equals(value.ToString()));
+        }
+
+        private static ICollection<string> GetMatches(string json, string key)
+        {
+            var regex = new Regex($"(?<=\"{key}\"\\ *:\\ *\")(?:\\\\\"|[^\"])*");
```
Comment (arcuri82) [inline client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs:30]: explain the regex in a comment

furthermore, are there no common libraries in .Net that do this kind of JSON handling? wondering if there is any edge case this regex would not handle

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 4]
```diff
@@ -399,12 +437,49 @@ class TestCaseWriter {
...
+                    )
                 }
             }
         }
-        lines.append(")")
+
+        if (format.isCsharp()) {
+            if (hasBody) {
+                when (verb) {
+                    "post", "put" -> lines.append(", httpContent);")
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:467]: what about patch?   better to refactor in a private method called `isVerbWithPossibleBodyPayload(verb: String)`

**Review by arcuri82 — CHANGES_REQUESTED** [review]

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs (hunk 2 above)
Comment (amidgol) [reply client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs]: This class is removed in the latest push

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs (hunk 1 above)
Comment (amidgol) [reply client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs]: This class is removed in the latest push

File: client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs (hunk 3 above)
Comment (amidgol) [reply client-dotnet/src/EvoMaster.Client.Util/JsonInspector.cs]: This class is removed in the latest push

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt (hunk 4 above)
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Done

**Review by arcuri82 — CHANGES_REQUESTED** [review]

**Comment by arcuri82** [comment]
Hi! thx ;)
CI is failing. please fix for failing tests.

Furthermore, looks like there were a few issue with merging of `TestCaseWriter.kt`

File: client-dotnet/src/EvoMaster.Controller/Controllers/db/Containers/DockerContainerBase.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Docker.DotNet;
+using Docker.DotNet.Models;
+
+namespace EvoMaster.Controller.Controllers.db
```
[the remark this answers is not in the document]
Comment (amidgol) [reply client-dotnet/src/EvoMaster.Controller/Controllers/db/Containers/DockerContainerBase.cs:7]: Done, also added for DatabaseStarter class

File: core/src/main/kotlin/org/evomaster/core/Main.kt [hunk 6]
```diff
@@ -94,7 +94,7 @@ class Main {
                                 "\n  Make sure the EvoMaster Driver for the system under test is running correctly.")
 
                     is SutProblemException ->
-                        logError("ERROR in the Remote EvoMaster Driver: ${cause.message}" +
+                        logError("ERROR                      in the Remote EvoMaster Driver: ${cause.message}" +
                                 "\n  Look at the logs of the EvoMaster Driver to help debugging this problem.")
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/Main.kt]: No idea how this has happened but fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 7]
```diff
@@ -39,20 +42,15 @@ class TestCaseWriter {
     //TODO: refactor in constructor, and take out of convertToCompilableTestCode
     private var format: OutputFormat = OutputFormat.JAVA_JUNIT_4
     private lateinit var configuration: EMConfig
-
-    /*
-        The following are just for REST
-        TODO: refactor, considering we are adding GraphQL and others
-     */
     private lateinit var expectationsWriter: ExpectationsWriter
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: These were removed on merge. Now they are back

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 8]
```diff
@@ -39,20 +42,15 @@ class TestCaseWriter {
...
 
     companion object {
         private val log = LoggerFactory.getLogger(TestCaseWriter::class.java)
 
-        /**
-         *   Internal flag to mark cases which do not support yet
+        /*
+            Internal flag to mark cases which do not support yet
          */
         const val NOT_COVERED_YET = "NotCoveredYet"
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Thanks for pointing out. Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 9]
```diff
@@ -123,19 +130,12 @@ class TestCaseWriter {
...
-                    SqlWriter.handleDbInitialization(format, ind.individual.seeInitializingActions(), lines)
-                }
-            } else {
-                if (ind.individual is GraphQLIndividual) {
-                    //TODO refactor
-                    if (ind.individual.dbInitialization.isNotEmpty()) {
-                        SqlWriter.handleDbInitialization(format, ind.individual.dbInitialization, lines)
-                    }
+                if (ind.individual.dbInitialization.isNotEmpty()) {
+                    SqlWriter.handleDbInitialization(format, ind.individual.dbInitialization, lines)
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 10]
```diff
@@ -123,19 +130,12 @@ class TestCaseWriter {
...
                 }
             }
+
             if (test.hasChainedLocations()) {
-                assert(ind.individual is RestIndividual)
                 /*
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:156]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 11]
```diff
@@ -147,44 +147,28 @@ class TestCaseWriter {
...
-                }
-            }
-            }
-            if (test.test.individual is GraphQLIndividual) {
-                test.test.evaluatedActions().forEach { a ->
-                    handleGraphQlCall(a, lines, baseUrlOfSut)
+            test.test.evaluatedActions().forEach { a ->
+                when (a.action) {
+                    is RestCallAction -> handleRestCall(a, lines, baseUrlOfSut)
+                    else -> throw IllegalStateException("Cannot handle " + a.action.getName())
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 12]
```diff
@@ -193,14 +177,8 @@ class TestCaseWriter {
...
         }
-        return lines
-    }
 
-    private fun handleEvaluatedAction(a: EvaluatedAction, lines: Lines, baseUrlOfSut: String) {
-        when (a.action) {
-            is RestCallAction -> handleRestCall(a, lines, baseUrlOfSut)
-            else -> throw IllegalStateException("Cannot handle " + a.action.getName())
-        }
+        return lines
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 13]
```diff
@@ -212,53 +190,37 @@ class TestCaseWriter {
...
         } else {
             addRestCallLines(call, lines, res, baseUrlOfSut)
         }
     }
 
-    private fun handleGraphQlCall(
-            evaluatedAction: EvaluatedAction,
-            lines: Lines,
-            baseUrlOfSut: String
+    private fun addRestCallInTryCatch(
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 14]
```diff
@@ -283,51 +245,7 @@ class TestCaseWriter {
...
-                     */
-                    lines.add("fail(\"Expected exception\");")
-                }
-            }
-        }
-
-        when {
-            format.isJavaOrKotlin() -> lines.add("} catch(Exception e){")
-            format.isJavaScript() -> lines.add("} catch(e){")
+            format.isCsharp() -> lines.add("} catch(Exception e){")
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:304]: Looks like fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 15]
```diff
@@ -381,34 +306,6 @@ class TestCaseWriter {
         }
     }
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 16]
```diff
@@ -434,29 +331,14 @@ class TestCaseWriter {
...
+            format.isCsharp() -> lines.append("Client.DefaultRequestHeaders.Clear();\n")
         }
         //TODO: check for C#
-        if (!format.isJavaScript() && !format.isCsharp()) {
-
-            lines.append(getAcceptHeader(call, res))
-        }
-    }
-
-    private fun handleGQLFirstLine(call: GraphQLAction, lines: Lines, res: GraphQlCallResult) {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 17]
```diff
@@ -502,30 +389,34 @@ class TestCaseWriter {
...
+                val extract =
+                    "$resVarName.extract().body().path$extraTypeInfo(\"${res.getResourceIdName()}\").toString()"
 
                 lines.add("${locationVar(call.path.lastElement())} = \"$baseUri/\" + $extract")
                 lines.appendSemicolon(format)
             }
         }
     }
 
-    private fun handleGQLLastLine(lines: Lines) {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:513]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 18]
```diff
@@ -546,28 +437,50 @@ class TestCaseWriter {
                 lines.append("$path?\" + ")
 
                 lines.indented {
-                    (0 until elements.lastIndex).forEach { i -> lines.add("\"${GeneUtils.applyEscapes(elements[i], mode = GeneUtils.EscapeMode.SQL, format = format)}&\" + ") }
-                    lines.add("\"${GeneUtils.applyEscapes(elements.last(), mode = GeneUtils.EscapeMode.SQL, format = format)}\"")
+                    (0 until elements.lastIndex).forEach { i ->
+                        lines.add(
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: This is due to reformatting code by pressing Ctrl+Alt+L

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 19]
```diff
@@ -592,31 +506,8 @@ class TestCaseWriter {
                 handleResponseContents(lines, res)
             }
         } else if (partialOracles.generatesExpectation(call, res)
-                && format.isJavaOrKotlin()) lines.add(".then()")
-    }
-
-    private fun handleGQLResponse(call: GraphQLAction, res: GraphQlCallResult, lines: Lines) {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:598]: This is due to reformatting code by pressing Ctrl+Alt+L in intellij

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 20]
```diff
@@ -625,10 +516,26 @@ class TestCaseWriter {
         } else {
             when (resContentsItem::class) {
                 Double::class -> return "numberMatches(${resContentsItem as Double})"
-                String::class -> return "containsString(\"${GeneUtils.applyEscapes(resContentsItem as String, mode = GeneUtils.EscapeMode.ASSERTION, format = format)}\")"
+                String::class -> return "containsString(\"${
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:683]: This is due to reformatting code by pressing Ctrl+Alt+L

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 21]
```diff
@@ -546,28 +437,50 @@ class TestCaseWriter {
...
+                    lines.append(");")
+            }
+            lines.add("responseBody = await response.Content.ReadAsStringAsync();")
+        } else
+            lines.append(")")
+    }
 
-        if (format.isKotlin()) {
-            lines.append("\"\${$baseUrlOfSut}")
-        } else {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 22]
```diff
@@ -730,98 +638,27 @@ class TestCaseWriter {
 
                             bodyString.isNullOrBlank() -> lines.add(".body(isEmptyOrNullString())")
 
-                            else -> lines.add(".body(containsString(\"${
-                            GeneUtils.applyEscapes(bodyString, mode = GeneUtils.EscapeMode.BODY, format = format)
-                            }\"))")
+                            else -> lines.add(
+                                ".body(containsString(\"${
+                                    GeneUtils.applyEscapes(
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:807]: Again this is due to reformatting code

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 23]
```diff
@@ -730,98 +638,27 @@ class TestCaseWriter {
...
                 } else {
-                    lines.add(".body(containsString(\"${
-                    GeneUtils.applyEscapes(bodyString, mode = GeneUtils.EscapeMode.TEXT, format = format)
-                    }\"))")
-                }
-            }
-        }
-    }
-
-    private fun handleGQLResponseContents(lines: Lines, res: GraphQlCallResult) {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:751]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 24]
```diff
@@ -945,56 +831,19 @@ class TestCaseWriter {
             when {
                 format.isJavaOrKotlin() -> lines.add(".contentType(\"application/x-www-form-urlencoded\")")
                 format.isJavaScript() -> lines.add(".set('Content-Type','application/x-www-form-urlencoded')")
+                format.isCsharp() -> lines.add("Client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(\"application/x-www-form-urlencoded\"));")
             }
-            lines.add(".$send(\"$form\")")
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:949]: In fact this is not removed, it is only put after an if statement

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 25]
```diff
@@ -1012,39 +861,10 @@ class TestCaseWriter {
...
-                format.isJavaOrKotlin() -> lines.add(".cookies(${CookieWriter.cookiesName(cookieLogin)})")
-                format.isJavaScript() -> lines.add(".set('Cookies', ${CookieWriter.cookiesName(cookieLogin)})")
+            .filter { !prechosenAuthHeaders.contains(it.name) }
+            .forEach {
+                lines.add(".$set(\"${it.name}\", ${it.gene.getValueAsPrintableString(targetFormat = format)})")
             }
-        }
-    }
-
-    private fun handleGQLHeaders(call: GraphQLAction, lines: Lines) {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:1029]: Fixed

File: core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt [hunk 26]
```diff
@@ -1067,39 +887,28 @@ class TestCaseWriter {
...
         val accepted = call.produces.filter { res.getBodyType().toString().contains(it, true) }
 
-        if (accepted.size == 1)
-            return "$accept\"${accepted.first()}\")"
-        else
-        //FIXME: there seems to have been something or a problem
-            return "$accept\"*/*\")"
-    }
-
-    private fun getAcceptGQLHeader(call: GraphQLAction, res: GraphQlCallResult): String {
```
[the remark this answers is not in the document]
Comment (amidgol) [reply core/src/main/kotlin/org/evomaster/core/output/TestCaseWriter.kt:1087]: Fixed

**Pull request merged**